File: apcs/202109/g278. 4. 美食博覽會/main.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main(){
    int n, k, x, ans = 0;
    scanf("%d%d", &n, &k);
    int DP0 = 0;
    int DP[n][k+1];
    int last[1000000];
    for (int i=0; i<1000000; i++)
        last[i] = -1;
    
    for (int i=1; i<=n; i++) {
        scanf("%d", &x);
        DP0 = min(DP0+1, i-last[x]);
        last[x] = i;
        for (int cur_k=1; cur_k<=k; cur_k++) {
            DP[i][cur_k] = max(
            DP[i-1][cur_k],
            DP[i-DP0][cur_k-1] + DP0);
            ans = max(ans, DP[i][k]);
        }
    }
    printf("%d\n", ans);
}

File: apcs/202109/g278. 4. 美食博覽會/main.py
import sys
from collections import defaultdict
n, k = map(int, input().split())
A = list(map(int, sys.stdin.readline().split()))
last = defaultdict(lambda: -1)
ans, DP0 = 0, 0
DP = [[0]*(k+1) for _ in range(n)]

for cur_n, x in enumerate(A):
    DP0 = min(DP0+1, cur_n-last[x])
    last[x] = cur_n
    for cur_k in range(1, k+1):
        DP[cur_n][cur_k] = max(
            DP[cur_n-1][cur_k],
            DP[cur_n-DP0][cur_k-1]+DP0
        )
        ans = max(ans, DP[cur_n][cur_k])
print(ans)